subprojects/core/src/main/java/org/gradle/api/internal/plugins/DefaultPluginRegistry.java

```diff
@@ -105,6 +106,35 @@ public PluginRegistry createChild(final ClassLoaderScope lookupScope) {
         return null;
     }
 
+    @Override
+    public Optional<PluginId> findPluginForClass(Class<?> clazz) {
+        if (parent != null) {
+            Optional<PluginId> result = parent.findPluginForClass(clazz);
+            if (result.isPresent()) {
+                return result;
+            }
+        }
+
+        Optional<PluginImplementation<?>> impl = Optional.ofNullable(Cast.uncheckedCast(uncheckedGet(classMappings, clazz)));
+        if (impl.isPresent()) {
```

Comment:
Would it make sense to use `impl.map()` or `impl.flatMap()`?

Reply:
We could but because of the inner loop I find the code more understandable as is (otherwise we need 2 levels of `map`).